gateway/bucketfs.go

```diff
@@ -0,0 +1,85 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/textileio/textile/api/client"
+)
+
+type serveBucketFileSystem interface {
+	Exists(bucket, pth string) bool
+	Write(bucket, pth string, writer io.Writer) error
+	ValidHost() string
+}
+
+type bucketFileSystem struct {
+	client  *client.Client
+	auth    client.Auth
+	timeout time.Duration
+	host    string
+}
+
+// serveBucket returns a middleware handler that serves files in a bucket.
+func serveBucket(fs serveBucketFileSystem) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		bucket, err := bucketFromHost(c.Request.Host, fs.ValidHost())
+		if err != nil {
+			return
+		}
+		if fs.Exists(bucket, c.Request.URL.Path) {
+			c.Writer.WriteHeader(http.StatusOK)
+			ctype := mime.TypeByExtension(filepath.Ext(c.Request.URL.Path))
```

Comment:
Would be nice to use more advanced content type detection, but we'd need the client to implement a `ReadSeeker`.